### PR TITLE
receiver: Split - Upstream the changes from the PR #3580

### PR DESCRIPTION
**Pull request by yashrsharma44**
Rebased #3580 with the `main` branch.

> Notes:
> * There were few issues with the dynamic type checking for error, I am getting the following error - 
> `errs, ok := unwrappedErr.(errutil.MultiError)`
> `unwrappedErr (variable of type error) cannot have dynamic type errutil.MultiError (missing method Error)`

Fixed it by creating another method for multierror - `Error()`. Let me know if the above makes sense.
Would love some reviews on this.
cc @brancz

### CODE REVIEW

File: pkg/receive/multitsdb.go [hunk 1]
```diff
@@ -44,8 +43,6 @@ type MultiTSDB struct {
 	allowOutOfOrderUpload bool
 }
 
-// NewMultiTSDB creates new MultiTSDB.
-// NOTE: Passed labels has to be sorted by name.
```
Comment (yeya24) [inline pkg/receive/multitsdb.go:48]: Would be good to keep the comments?

File: pkg/receive/multitsdb.go [hunk 2]
```diff
@@ -265,7 +262,7 @@ func (t *MultiTSDB) TSDBStores() map[string]storepb.StoreServer {
 
 func (t *MultiTSDB) startTSDB(logger log.Logger, tenantID string, tenant *tenant) error {
 	reg := prometheus.WrapRegistererWith(prometheus.Labels{"tenant": tenantID}, t.reg)
-	lset := labelpb.ExtendSortedLabels(t.labels, labels.FromStrings(t.tenantLabelName, tenantID))
+	lbls := append(t.labels, labels.Label{Name: t.tenantLabelName, Value: tenantID})
```
Comment (yeya24) [inline pkg/receive/multitsdb.go:265]: I guess we still want to have `labelpb.ExtendSortedLabels` here

File: test/e2e/e2ethanos/services.go [hunk 3]
```diff
@@ -259,30 +248,63 @@ func NewReceiverWithConfigWatcher(sharedDir string, networkName string, name str
...
 		9091,
-		8081,
 	)
 	receiver.SetUser(strconv.Itoa(os.Getuid()))
 	receiver.SetBackoff(defaultBackoffConfig)
-
 	return receiver, nil
 }
 
+func NewRouter(sharedDir string, networkName string, name string, replicationFactor int, hashring ...route.HashringConfig) (*e2e.HTTPService, error) {
```
Comment (yeya24) [inline test/e2e/e2ethanos/services.go:268]: Let's call it `NewReceiveRouter`?

File: test/e2e/receive_test.go [hunk 4]
```diff
@@ -70,35 +70,33 @@ func TestReceive(t *testing.T) {
...
-		r2, err = e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "2", 1, h)
+		// Create a router.
+		route1, err := e2ethanos.NewRouter(s.SharedDir(), s.NetworkName(), "1", 1, h)
 		testutil.Ok(t, err)
-		r3, err = e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "3", 1, h)
-		testutil.Ok(t, err)
-		testutil.Ok(t, s.StartAndWaitReady(r1, r2, r3))
+		x := s.StartAndWaitReady(r1, r2, r3, route1)
+		println(x)
+		time.Sleep(5 * time.Minute)
```
Comment (yeya24) [inline test/e2e/receive_test.go:93]: Why do we need to sleep here?

File: test/e2e/e2ethanos/services.go [hunk 5]
```diff
@@ -188,39 +188,28 @@ func NewQuerier(sharedDir, name string, storeAddresses, fileSDStoreAddresses, ru
...
-			"--tsdb.path":                  filepath.Join(container, "data"),
-			"--log.level":                  infoLogLevel,
-			"--receive.replication-factor": strconv.Itoa(replicationFactor),
-			"--receive.local-endpoint":     localEndpoint,
-			"--receive.hashrings":          string(b),
+			"--debug.name":           fmt.Sprintf("receive-%v", name),
+			"--grpc-address":         ":9091",
+			"--grpc-grace-period":    "0s",
+			"--http-address":         ":8080",
+			"--remote-write.address": ":8081",
```
Comment (yeya24) [inline test/e2e/e2ethanos/services.go:209]: This flag is already removed.

File: test/e2e/receive_test.go (hunk 4 above)
Comment (yashrsharma44) [reply test/e2e/receive_test.go:93]: Seems like this was written to debug, removing this

File: test/e2e/e2ethanos/services.go (hunk 3 above)
Comment (yashrsharma44) [reply test/e2e/e2ethanos/services.go:268]: Sure

File: pkg/receive/multitsdb.go (hunk 1 above)
Comment (yashrsharma44) [reply pkg/receive/multitsdb.go]: Sure

File: test/e2e/e2ethanos/services.go (hunk 3 above)
Comment (yashrsharma44) [reply test/e2e/e2ethanos/services.go]: Yup 👍

File: pkg/receive/multitsdb.go (hunk 2 above)
Comment (yashrsharma44) [reply pkg/receive/multitsdb.go]: Okay, reverting back

File: test/e2e/e2ethanos/services.go (hunk 5 above)
Comment (yashrsharma44) [reply test/e2e/e2ethanos/services.go]: Thanks, I was not able to debug this issue, thanks for pointing out!

**Comment by yashrsharma44** [comment]
I have updated the changes, take a look whenever free 🤗 @yeya24

File: test/e2e/e2ethanos/services.go [hunk 6]
```diff
@@ -259,30 +247,63 @@ func NewReceiverWithConfigWatcher(sharedDir string, networkName string, name str
...
+	router := e2e.NewHTTPService(
+		fmt.Sprintf("receive-route-%v", name),
+		DefaultImage(),
+		// TODO(bwplotka): BuildArgs should be interface.
+		e2e.NewCommand("receive-route", e2e.BuildArgs(map[string]string{
+			"--debug.name":                 fmt.Sprintf("receive-route-%v", name),
+			"--http-address":               ":8080",
+			"--remote-write.address":       ":8081",
+			"--label":                      fmt.Sprintf(`receive="%s"`, name),
+			"--tsdb.path":                  filepath.Join(container, "data"),
```
Comment (yeya24) [inline test/e2e/e2ethanos/services.go:292]: Looks like the receive router doesn't have `--label`  and `--tsdb.path` flags

File: cmd/thanos/receive.go [hunk 7]
```diff
@@ -472,58 +221,36 @@ func runReceive(
...
+				grpcserver.WithGracePeriod(grpcGracePeriod),
+				grpcserver.WithTLSConfig(tlsCfg),
+			)
+
+			level.Info(logger).Log("msg", "listening for StoreAPI and WritableStoreAPI gRPC", "address", grpcBindAddr)
+			statusProber.Ready()
+			if err := s.ListenAndServe(); err != nil {
+				return errors.Wrap(err, "serve gRPC")
 			}
+			statusProber.Ready()
```
Comment (yeya24) [inline cmd/thanos/receive.go:247]: This line is the same as line 243

File: cmd/thanos/receive_route.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+		g.Add(func() error {
+			for {
+				select {
+				case h, ok := <-updates:
+					if !ok {
+						return nil
+					}
+					webHandler.Hashring(h)
+					msg := "hashring has changed; server is not ready to receive web requests"
+					statusProber.NotReady(errors.New(msg))
```
Comment (yeya24) [inline cmd/thanos/receive_route.go:178]: We never set the probe to ready in this code. Seems that's why E2E test says that the receive router is not ready

File: test/e2e/e2ethanos/services.go (hunk 6 above)
Comment (yashrsharma44) [reply test/e2e/e2ethanos/services.go:292]: Thanks for pointing out!

File: cmd/thanos/receive.go (hunk 7 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go:247]: Okay

File: cmd/thanos/receive_route.go (hunk 8 above)
Comment (yashrsharma44) [reply cmd/thanos/receive_route.go:178]: Don't we set up the probe ready in [#L198](https://github.com/thanos-io/thanos/pull/3845/files#diff-5dd9b8ab9e76f0f359092fc4244dc009a8c7478e000c730cf969a49dcf9fa7a9R196) ?

Comment (yeya24) [reply cmd/thanos/receive_route.go:178]: Healthy and ready are two different states. CC @kakkoyun for double check.

Comment (yashrsharma44) [reply cmd/thanos/receive_route.go:178]: My bad, we are not setting prober ready out in this code

Comment (yashrsharma44) [reply cmd/thanos/receive_route.go:178]: Any ideas, where should we mark the receive ready?

**Comment by yashrsharma44** [comment]
I wonder if this suggestion is helpful in `make lint` - 
```powershell
/home/yash/golib/src/github.com/thanos-io/thanos/pkg/route/handler_test.go:234:31: declaration "NewCounterVec" from package "github.com/prometheus/client_golang/prometheus" shouldn't be used, suggested: "github.com/prometheus/client_golang/prometheus/promauto.{NewCounter,NewCounterVec,NewCounterVec,NewGauge,NewGaugeVec,NewGaugeFunc,NewHistorgram,NewHistogramVec,NewSummary,NewSummaryVec}"
/home/yash/golib/src/github.com/thanos-io/thanos/pkg/route/handler_test.go:240:28: declaration "NewCounterVec" from package "github.com/prometheus/client_golang/prometheus" shouldn't be used, suggested: "github.com/prometheus/client_golang/prometheus/promauto.{NewCounter,NewCounterVec,NewCounterVec,NewGauge,NewGaugeVec,NewGaugeFunc,NewHistorgram,NewHistogramVec,NewSummary,NewSummaryVec}"
/home/yash/golib/src/github.com/thanos-io/thanos/pkg/route/handler_test.go:246:33: declaration "NewGauge" from package "github.com/prometheus/client_golang/prometheus" shouldn't be used, suggested: "github.com/prometheus/client_golang/prometheus/promauto.{NewCounter,NewCounterVec,NewCounterVec,NewGauge,NewGaugeVec,NewGaugeFunc,NewHistorgram,NewHistogramVec,NewSummary,NewSummaryVec}"
```
because a similar issue was reported here after applying the suggestion - https://github.com/prometheus/client_golang/issues/716

**Comment by yashrsharma44** [comment]
Any idea to get the tests fixed? 🤔 cc @brancz @kakkoyun

**Comment by brancz** [comment]
I don't have time for this right now, but what I'm a bit confused about is that

1) There seem to be commits missing compared to the other PR
2) The e2e tests were passing on the last state of the other PR

**Comment by yeya24** [comment]
I will do some investigation today.

File: cmd/thanos/receive_route.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+				select {
+				case h, ok := <-updates:
+					if !ok {
+						return nil
+					}
+					msg := "hashring has changed; server is not ready to receive web requests"
+					statusProber.NotReady(errors.New(msg))
+					level.Info(logger).Log("msg", msg)
+					webHandler.Hashring(h)
+					statusProber.Ready()
```
Comment (yashrsharma44) [inline cmd/thanos/receive_route.go:179]: Added the ready prober here,
Need to check if the change makes sense 🤔 cc @squat

**Comment by yeya24** [comment]
Can you please fix the Circle CI test and linting tests?

File: cmd/thanos/receive_route.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+	}
+
+	level.Debug(logger).Log("msg", "setting up http server")
+	{
+		srv := httpserver.New(logger, reg, comp, httpProbe,
+			httpserver.WithListen(httpBindAddr),
+			httpserver.WithGracePeriod(httpGracePeriod),
+		)
+		g.Add(func() error {
+			statusProber.Ready()
```
Comment (kakkoyun) [inline cmd/thanos/receive_route.go:197]:
```suggestion
			statusProber.Healthy()
```

File: cmd/thanos/receive_route.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+			defer statusProber.Healthy()
```
Comment (kakkoyun) [inline cmd/thanos/receive_route.go:198]:
```suggestion
```

File: cmd/thanos/main.go [hunk 12]
```diff
@@ -29,10 +28,6 @@ import (
 )
 
 func main() {
-	// We use mmaped resources in most of the components so hardcode PanicOnFault to true. This allows us to recover (if we can e.g if queries
```
Comment (kakkoyun) [inline cmd/thanos/main.go:32]: Why do we remove this? Is this intentional?

cc @yashrsharma44

File: pkg/errutil/multierror.go [hunk 13]
```diff
@@ -32,6 +32,11 @@ func (es MultiError) Err() error {
 	return NonNilMultiError(es)
 }
 
+// Error implements the error interface.
+func (es MultiError) Error() string {
```
Comment (kakkoyun) [inline pkg/errutil/multierror.go:36]: Let's check the recent error handling changes

https://github.com/thanos-io/thanos/pull/3833
https://github.com/thanos-io/thanos/pull/3836

@yashrsharma44

File: pkg/route/config_test.go [hunk 14]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) The Thanos Authors.
 // Licensed under the Apache License 2.0.
 
-package receive
+package route
```
Comment (kakkoyun) [inline pkg/route/config_test.go:4]: I think this should go under `receive`. This doesn't seem generic enough.

File: pkg/route/error.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (a expectedErrors) Len() int           { return len(a) }
+func (a expectedErrors) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a expectedErrors) Less(i, j int) bool { return a[i].count < a[j].count }
+
+// determineWriteErrorCause extracts a sentinel error that has occurred more than the given threshold from a given fan-out error.
+// It will inspect the error's cause if the error is a MultiError,
+// It will return cause of each contained error but will not traverse any deeper.
+func determineWriteErrorCause(err error, threshold int) error {
```
Comment (kakkoyun) [inline pkg/route/error.go:35]: Let's make sure we have parity https://github.com/thanos-io/thanos/blob/befb0257e8a3d49cb80b00794a4892abfb7ccb72/pkg/receive/handler.go#L689

**Review by kakkoyun — COMMENTED** [review]
Either we're missing some context on this or we miss certain parts of the code 😅

File: cmd/thanos/receive_route.go (hunk 9 above)
Comment (kakkoyun) [reply cmd/thanos/receive_route.go:179]: I'm sure this is correct.
It seems like I miss the part that connects hasring updates to TSDB flush/open.

File: pkg/route/hashring.go [hunk 16]
```diff
@@ -29,25 +28,25 @@ func (i *insufficientNodesError) Error() string {
...
+// Hashring finds the correct node to handle a given label-set
 // for a specified tenant.
 // It returns the node and any error encountered.
 type Hashring interface {
-	// Get returns the first node that should handle the given tenant and time series.
-	Get(tenant string, timeSeries *prompb.TimeSeries) (string, error)
-	// GetN returns the nth node that should handle the given tenant and time series.
-	GetN(tenant string, timeSeries *prompb.TimeSeries, n uint64) (string, error)
+	// Get returns the first node that should handle the given tenant and label-set.
+	Get(tenant string, labelSet []labelpb.ZLabel) (string, error)
```
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/route/hashring.go:36]: Yeah this is unfortunate as this work was rebased a number of times in between.

File: cmd/thanos/main.go (hunk 12 above)
Comment (squat) [reply cmd/thanos/main.go:32]: this seems like it was left-over from testing. Should be reverted

File: cmd/thanos/main.go [hunk 17]
```diff
@@ -48,7 +53,6 @@ func main() {
 	registerCompact(app)
 	registerTools(app)
 	registerReceive(app)
-	registerReceiveRoute(app)
```
Comment (yeya24) [inline cmd/thanos/main.go]: We still need this

Comment (yashrsharma44) [reply cmd/thanos/main.go]: My bad, I kind of reverted all the changes of `main.go`

File: cmd/thanos/receive_route.go (hunk 8 above)
Comment (yashrsharma44) [reply cmd/thanos/receive_route.go:181]: Resolved using https://github.com/thanos-io/thanos/pull/3845#discussion_r586113729

File: cmd/thanos/receive.go [hunk 18]
```diff
@@ -285,180 +183,33 @@ func runReceive(
...
-					statusProber.NotReady(errors.New(msg))
-					level.Info(logger).Log("msg", msg)
-					hashringChangedChan <- struct{}{}
-				case <-cancel:
-					return nil
-				}
-			}
+			statusProber.Healthy()
+			err := srv.ListenAndServe()
+			return err
```
[the remark this answers is not in the document]
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Done

File: cmd/thanos/receive_route.go (hunk 10 above)
Comment (yashrsharma44) [reply cmd/thanos/receive_route.go]: Done

File: cmd/thanos/receive_route.go (hunk 11 above)
Comment (yashrsharma44) [reply cmd/thanos/receive_route.go]: Done

File: cmd/thanos/receive.go [hunk 19]
```diff
@@ -557,7 +285,7 @@ func runReceive(
 
 				// Before quitting, ensure all blocks are uploaded.
 				defer func() {
-					<-uploadC // Closed by storage routine when it's done.
+
```
[the remark this answers is not in the document]
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Done

File: pkg/route/hashring.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yashrsharma44) [reply pkg/route/hashring.go]: Reverted back

File: pkg/route/config_test.go (hunk 14 above)
Comment (yashrsharma44) [reply pkg/route/config_test.go:4]: Let's rename the package to `receive-route`?

File: pkg/route/error.go (hunk 15 above)
Comment (yashrsharma44) [reply pkg/route/error.go:35]: Can you elaborate on this?

File: cmd/thanos/receive.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Done

File: pkg/route/hashring.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yashrsharma44) [reply pkg/route/hashring.go]: Reverted changes

**Comment by yashrsharma44** [comment]
Fixed the failing tests as we were using `MultiError` type assertion instead of `NonNilMultiError` assertion. Now the tests are passing. 
cc @kakkoyun @squat can you review if the current changes make sense in accordance to the `main` branch?

File: pkg/route/config_test.go (hunk 14 above)
Comment (yashrsharma44) [reply pkg/route/config_test.go]: Done

File: pkg/errutil/multierror.go (hunk 13 above)
Comment (yashrsharma44) [reply pkg/errutil/multierror.go]: Done, have used the new Error changes 👍

File: pkg/receive_route/hashring.go [hunk 20]
```diff
@@ -182,18 +182,3 @@ func HashringFromConfigWatcher(ctx context.Context, updates chan<- Hashring, cw
 		}
 	}
 }
-
```
Comment (kakkoyun) [inline pkg/receive_route/hashring.go:185]: This is a regression I believe. This functionality added by a recent PR https://github.com/thanos-io/thanos/pull/3121
@yashrsharma44

File: docs/components/receive.md [hunk 21]
```diff
@@ -108,29 +108,6 @@ Flags:
                                  TLS CA to verify clients against. If no client
                                  CA is specified, there is no client
                                  verification on server side. (tls.NoClientCert)
-      --remote-write.address="0.0.0.0:19291"
```
Comment (kakkoyun) [inline docs/components/receive.md:111]: Yes, we remove these however, where are we adding those? We need to generate documentation for router as well.

**Review by kakkoyun — CHANGES_REQUESTED** [review]
I have just skimmed again and pointed out the major things. I'll review it in depth when I have time.

We should be super considered about merging this. @yashrsharma44 Please check all the subsequent PRs after this branched is created. And also we should have all the necessary facilities for the new subcommand. Moreover I don't see any changes on alerts or dashboard don't we have any changes regarding those parts?

I understand this is just a rebase effort but this could break receiver on master. If there are things missing in the previous PR, these should be documented by an issue or a document so that we can fix all those before the next major release. cc @Chans321

**Comment by yashrsharma44** [comment]
TODO:
- [ ] Traverse PRs and add anything missing to this PR. - Need to be taken care in this PR
- [ ] Make sure documentation for the new command exists. - Enable autogeneration of documentation for receive-router
- [ ] Make sure dashboards and alerts work with the new command. - We can create an issue to track the shortcomings related to dashboards and alerts

File: pkg/receive_route/hashring.go (hunk 20 above)
Comment (yashrsharma44) [reply pkg/receive_route/hashring.go]: Have updated the changes, thanks for identifying this :+1:

**Comment by bwplotka** [comment]
As promised, the modification of the proposal is ready: https://github.com/thanos-io/thanos/pull/4036 🤗  PTAL!

**Comment by kakkoyun** [comment]
Deprecated by https://github.com/thanos-io/thanos/pull/4231

**Pull request closed without merging**